gtsam_unstable/geometry/Similarity3.cpp

```diff
@@ -85,10 +86,104 @@ Point3 Similarity3::transformFrom(const Point3& p, //
   return s_ * q;
 }
 
+Pose3 Similarity3::transformFrom(const Pose3& T) const {
+  Rot3 R = R_.compose(T.rotation());
+  Point3 t = Point3(s_ * (R_ * T.translation() + t_));
+  return Pose3(R, t);
+}
+
 Point3 Similarity3::operator*(const Point3& p) const {
   return transformFrom(p);
 }
 
+std::vector<Point3Pair> Similarity3::SubtractCentroids(const std::vector<Point3Pair>& abPointPairs, const Point3& aCentroid, const Point3& bCentroid) {
+  std::vector<Point3Pair> d_abPointPairs;
+  Point3 aPoint, bPoint;
+  for (const Point3Pair& abPair : abPointPairs) {
+    std::tie(aPoint, bPoint) = abPair;
+    Point3 da = aPoint - aCentroid;
+    Point3 db = bPoint - bCentroid;
+    d_abPointPairs.emplace_back(da, db);
+  }
+  return d_abPointPairs;
+}
+
+std::pair<double, double> Similarity3::GetXY(const std::vector<Point3Pair>& d_abPointPairs, const Rot3& aRb) {
+  double x = 0, y = 0;
+  Point3 da, db;
+  for (const Point3Pair& d_abPair : d_abPointPairs) {
+    std::tie(da, db) = d_abPair;
+    Vector3 Rdb = aRb * db;
+    y += da.transpose() * Rdb;
+    x += Rdb.transpose() * Rdb;
+  }
+  return std::make_pair(x, y);
+}
+
+Matrix3 Similarity3::GetH(const std::vector<Point3Pair>& d_abPointPairs) {
+  Matrix3 H = Z_3x3;
+  Point3 da, db;
+  for (const Point3Pair& d_abPair : d_abPointPairs) {
+    std::tie(da, db) = d_abPair;
+    H += da * db.transpose();
+  }
+  return H;
+}
+
+Similarity3 Similarity3::AlignGivenRandCentroids(const std::vector<Point3Pair>& d_abPointPairs, const Rot3& aRb, const Point3& aCentroid, const Point3& bCentroid) {
+  double x, y;
+  std::tie(x, y) = GetXY(d_abPointPairs, aRb);
+  const double s = y / x;
+  // calculate translation
+  const Point3 aTb = (aCentroid - s * (aRb * bCentroid)) / s;
+  return Similarity3(aRb, aTb, s);
+}
+
+Similarity3 Similarity3::AlignGivenR(const std::vector<Point3Pair>& abPointPairs, const Rot3& aRb) {
+  // Refer to: http://www5.informatik.uni-erlangen.de/Forschung/Publikationen/2005/Zinsser05-PSR.pdf Chapter 3
+  // calculate centroids
+  Point3 aCentroid, bCentroid;
+  std::tie(aCentroid, bCentroid) = mean(abPointPairs);
+  // substract centroids
+  std::vector<Point3Pair> d_abPointPairs = SubtractCentroids(abPointPairs, aCentroid, bCentroid);
+  return AlignGivenRandCentroids(d_abPointPairs, aRb, aCentroid, bCentroid);
+}
+
+Similarity3 Similarity3::Align(const std::vector<Point3Pair>& abPointPairs) {
+  // Refer to: http://www5.informatik.uni-erlangen.de/Forschung/Publikationen/2005/Zinsser05-PSR.pdf Chapter 3
+  const size_t n = abPointPairs.size();
+  if (n < 3) throw std::runtime_error("input should have at least 3 pairs of points");  // we need at least three pairs
```

Comment:
Get rid of comment and intermediate variable `n`. Make sure to cppformat code you edited (not other code!) With Google style.